File: patterns/03_observer/python/speculae.py
#!/usr/bin/env python3
"""
OBSERVER PATTERN — Roman Fire-Signal Network (Speculae)
"Una specula ardet, omnes vident" — One tower burns, all see it

Run: python3 speculae.py
"""

from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List


# ─── OBSERVER INTERFACE ──────────────────────────────────────
class SignalObserver(ABC):
    @abstractmethod
    def on_signal(self, message: str, urgency: int) -> None: ...

    @property
    @abstractmethod
    def observer_name(self) -> str: ...


# ─── CONCRETE OBSERVERS ──────────────────────────────────────
class RomanLegate(SignalObserver):
    def __init__(self, name: str, legion: str):
        self._name   = name
        self._legion = legion

    @property
    def observer_name(self) -> str:
        return self._name

    def on_signal(self, message: str, urgency: int) -> None:
        if urgency >= 4:
            print(f"  ⚔  LEGATE {self._name} ({self._legion}): "
                  f"FULL MOBILIZATION! '{message}'")
        elif urgency >= 2:
            print(f"  📜 LEGATE {self._name} ({self._legion}): "
                  f"Prepare forces. '{message}'")
        else:
            print(f"  📜 LEGATE {self._name}: Noted. '{message}'")


class EmperorInRome(SignalObserver):
    @property
    def observer_name(self) -> str:
        return "Emperor Hadrian"

    def on_signal(self, message: str, urgency: int) -> None:
        if urgency >= 5:
            print(f"  👑 EMPEROR: BY JUPITER! SOUND ALL TRUMPETS! '{message}'")
        elif urgency >= 3:
            print(f"  👑 EMPEROR: Dispatch two legions immediately. '{message}'")
        else:
            print(f"  👑 EMPEROR: Send a response letter. '{message}'")


class ChroniclerAnnales(SignalObserver):
    def __init__(self):
        self._event_count = 0

    @property
    def observer_name(self) -> str:
        return "Chronicler Annales"

    def on_signal(self, message: str, urgency: int) -> None:
        self._event_count += 1
        print(f"  📖 CHRONICLER [Event #{self._event_count}]: "
              f"Recording — '{message}' (urgency={urgency})")

    @property
    def event_count(self) -> int:
        return self._event_count


class SupplyOfficer(SignalObserver):
    @property
    def observer_name(self) -> str:
        return "Supply Officer"

    def on_signal(self, message: str, urgency: int) -> None:
        if urgency >= 3:
            print(f"  🍞 SUPPLY OFFICER: Preparing emergency rations for mobilization.")
        # Low urgency signals ignored by supply


# ─── SUBJECT (SIGNAL TOWER) ──────────────────────────────────
class SignalTower:
    """
    The Subject — maintains a list of observers and notifies them all
    when fireSignal() is called.
    """

    def __init__(self, location: str):
        self._location  = location
        self._observers: List[SignalObserver] = []

    def subscribe(self, observer: SignalObserver) -> None:
        self._observers.append(observer)
        print(f"  + {observer.observer_name} subscribed to {self._location}")

    def unsubscribe(self, observer: SignalObserver) -> None:
        self._observers.remove(observer)
        print(f"  - {observer.observer_name} unsubscribed from {self._location}")

    def fire_signal(self, message: str, urgency: int = 1) -> None:
        print(f"\n🔥 SIGNAL from {self._location} "
              f"[URGENCY={urgency}/5]: {message}")
        print(f"   Notifying {len(self._observers)} observer(s):")
        for observer in self._observers:
            observer.on_signal(message, urgency)

    @property
    def observer_count(self) -> int:
        return len(self._observers)


# ─── DEMONSTRATION ───────────────────────────────────────────
def main() -> None:
    print("╔═══════════════════════════════════════════════╗")
    print("║   OBSERVER PATTERN — Python                   ║")
    print("║   Roman Fire-Signal Network (Speculae)        ║")
    print("╚═══════════════════════════════════════════════╝\n")

    tower = SignalTower("Hadrian's Wall — Milecastle 39")

    legate     = RomanLegate("Gnaeus Pompeius Julius", "Legio XX Valeria Victrix")
    emperor    = EmperorInRome()
    chronicler = ChroniclerAnnales()
    supply     = SupplyOfficer()

    print("── SUBSCRIBING OBSERVERS ───────────────────────")
    tower.subscribe(legate)
    tower.subscribe(emperor)
    tower.subscribe(chronicler)
    tower.subscribe(supply)

    print("\n── SIGNAL EVENTS ───────────────────────────────")
    tower.fire_signal("Scout reports: small Pictish hunting party", urgency=1)
    tower.fire_signal("Border patrol: 200 armed Picts approaching", urgency=3)
    tower.fire_signal("EMERGENCY: Full Pictish tribal invasion! 10,000 warriors!", urgency=5)

    print("\n── SUBSCRIPTION CHANGE ─────────────────────────")
    print("  (Emperor goes on holiday to Hadrian's Villa)")
    tower.unsubscribe(emperor)
    tower.fire_signal("Situation stabilised — Picts retreated", urgency=1)

    print("\n── SUMMARY ─────────────────────────────────────")
    print(f"  Total events chronicled: {chronicler.event_count}")
    print(f"  Current observers: {tower.observer_count}")
    print("  \"Una specula ardet, omnes vident!\"")


if __name__ == "__main__":
    main()

File: patterns/03_observer/php/speculae.php
<?php
/**
 * OBSERVER PATTERN — Roman Fire-Signal Network (Speculae)
 * "Una specula ardet, omnes vident" — One tower burns, all see it
 *
 * Run: php speculae.php
 */

declare(strict_types=1);

// ─── OBSERVER INTERFACE ──────────────────────────────────────
interface SignalObserver
{
    public function onSignal(string $message, int $urgency): void;
    public function getObserverName(): string;
}

// ─── CONCRETE OBSERVERS ──────────────────────────────────────
class RomanLegate implements SignalObserver
{
    public function __construct(
        private string $name,
        private string $legion
    ) {}

    public function getObserverName(): string { return $this->name; }

    public function onSignal(string $message, int $urgency): void
    {
        if ($urgency >= 4) {
            echo "  ⚔  LEGATE {$this->name} ({$this->legion}): "
               . "FULL MOBILIZATION! '{$message}'\n";
        } elseif ($urgency >= 2) {
            echo "  📜 LEGATE {$this->name} ({$this->legion}): "
               . "Prepare forces. '{$message}'\n";
        } else {
            echo "  📜 LEGATE {$this->name}: Noted. '{$message}'\n";
        }
    }
}

class EmperorInRome implements SignalObserver
{
    public function getObserverName(): string { return "Emperor Hadrian"; }

    public function onSignal(string $message, int $urgency): void
    {
        if ($urgency >= 5) {
            echo "  👑 EMPEROR: BY JUPITER! SOUND ALL TRUMPETS! '{$message}'\n";
        } elseif ($urgency >= 3) {
            echo "  👑 EMPEROR: Dispatch two legions immediately. '{$message}'\n";
        } else {
            echo "  👑 EMPEROR: Send a response letter. '{$message}'\n";
        }
    }
}

class ChroniclerAnnales implements SignalObserver
{
    private int $eventCount = 0;

    public function getObserverName(): string { return "Chronicler Annales"; }

    public function onSignal(string $message, int $urgency): void
    {
        ++$this->eventCount;
        echo "  📖 CHRONICLER [Event #{$this->eventCount}]: "
           . "Recording — '{$message}' (urgency={$urgency})\n";
    }

    public function getEventCount(): int { return $this->eventCount; }
}

class SupplyOfficer implements SignalObserver
{
    public function getObserverName(): string { return "Supply Officer"; }

    public function onSignal(string $message, int $urgency): void
    {
        if ($urgency >= 3) {
            echo "  🍞 SUPPLY OFFICER: Preparing emergency rations for mobilization.\n";
        }
    }
}

// ─── SUBJECT (SIGNAL TOWER) ──────────────────────────────────
class SignalTower
{
    /** @var SignalObserver[] */
    private array $observers = [];

    public function __construct(private string $location) {}

    public function subscribe(SignalObserver $observer): void
    {
        $this->observers[] = $observer;
        echo "  + " . $observer->getObserverName() . " subscribed to {$this->location}\n";
    }

    public function unsubscribe(SignalObserver $observer): void
    {
        $this->observers = array_filter(
            $this->observers,
            fn($o) => $o !== $observer
        );
        echo "  - " . $observer->getObserverName() . " unsubscribed from {$this->location}\n";
    }

    public function fireSignal(string $message, int $urgency = 1): void
    {
        $count = count($this->observers);
        echo "\n🔥 SIGNAL from {$this->location} [URGENCY={$urgency}/5]: {$message}\n";
        echo "   Notifying {$count} observer(s):\n";
        foreach ($this->observers as $observer) {
            $observer->onSignal($message, $urgency);
        }
    }

    public function getObserverCount(): int { return count($this->observers); }
}


// ─── DEMONSTRATION ───────────────────────────────────────────
echo "╔═══════════════════════════════════════════════╗\n";
echo "║   OBSERVER PATTERN — PHP                      ║\n";
echo "║   Roman Fire-Signal Network (Speculae)        ║\n";
echo "╚═══════════════════════════════════════════════╝\n\n";

$tower      = new SignalTower("Hadrian's Wall — Milecastle 39");
$legate     = new RomanLegate("Gnaeus Pompeius Julius", "Legio XX Valeria Victrix");
$emperor    = new EmperorInRome();
$chronicler = new ChroniclerAnnales();
$supply     = new SupplyOfficer();

echo "── SUBSCRIBING OBSERVERS ───────────────────────\n";
$tower->subscribe($legate);
$tower->subscribe($emperor);
$tower->subscribe($chronicler);
$tower->subscribe($supply);

echo "\n── SIGNAL EVENTS ───────────────────────────────";
$tower->fireSignal("Scout reports: small Pictish hunting party", 1);
$tower->fireSignal("Border patrol: 200 armed Picts approaching", 3);
$tower->fireSignal("EMERGENCY: Full Pictish tribal invasion! 10,000 warriors!", 5);

echo "\n── SUBSCRIPTION CHANGE ─────────────────────────\n";
echo "  (Emperor goes on holiday to Hadrian's Villa)\n";
$tower->unsubscribe($emperor);
$tower->fireSignal("Situation stabilised — Picts retreated", 1);

echo "\n── SUMMARY ─────────────────────────────────────\n";
echo "  Total events chronicled: " . $chronicler->getEventCount() . "\n";
echo "  Current observers: " . $tower->getObserverCount() . "\n";
echo "  \"Una specula ardet, omnes vident!\"\n";
